appender/src/test/java/com/vlkan/log4j2/redis/appender/RedisAppenderPubSubTest.java

```diff
@@ -0,0 +1,196 @@
+            CLASS_NAME,
+            redisAppenderName,
+            configBuilder -> configBuilder.add(configBuilder
+                .newAppender(redisAppenderName, "RedisAppender")
+                .addAttribute("host", redisHost)
+                .addAttribute("port", redisPort)
+                .addAttribute("username", redisUsername)
+                .addAttribute("password", redisPassword)
+                .addAttribute("key", redisKey)
+                .addAttribute("ignoreExceptions", false)
+                .addAttribute("usePubSubParadigm", true)
+                .add(configBuilder
+                    .newLayout("PatternLayout")
+                    .addAttribute("pattern", "%level %m"))
+                .addComponent(configBuilder
+                    .newComponent("RedisThrottlerConfig")
+                    // Batch size needs to be greater than 1, so that we can observe a partially filled batch push.
+                    .addAttribute("batchSize", 10)
+                    .addAttribute("bufferSize", 100)
+                    .addAttribute("flushPeriodMillis", 500L)
+                    .addAttribute("maxEventCountPerSecond", 35)
+                    .addAttribute("maxErrorCountPerSecond", 0)
+                    .addAttribute("maxByteCountPerSecond", 4194304))));
+
+    @Test
+    void appended_messages_should_be_persisted() {
+
+        // Create the logger.
+        LOGGER.debug("creating the logger");
+        Logger logger = loggerContextExtension
+            .getLoggerContext()
+            .getLogger(RedisAppenderPubSubTest.class.getCanonicalName());
+
+        // Create and log the messages.
+        int minMessageCount = 1;
+        int maxMessageCount = 100;
+        int expectedMessageCount = minMessageCount + RANDOM.nextInt(maxMessageCount - minMessageCount);
+
+        JedisPubSubTestImpl pubsub = new JedisPubSubTestImpl();
+        executor.execute(() -> redisClientExtension.getClient().subscribe(pubsub, redisKey));
+        LOGGER.debug("{} logging {} messages", LOGGER_PREFIX, expectedMessageCount);
+        RedisTestMessage[] expectedLogMessages = RedisTestMessage.createRandomArray(expectedMessageCount);
+        for (RedisTestMessage expectedLogMessage : expectedLogMessages) {
+            logger.log(expectedLogMessage.level, expectedLogMessage.message);
+        }
+
+        // Verify the logging.
+        verifyLogging(expectedLogMessages, expectedMessageCount, expectedMessageCount, pubsub);
+        executor.shutdown();
+    }
+
+    @Test
+    void throttler_should_not_flush_same_content_twice() {
+
+        // Create the logger.
+        LOGGER.debug("{} creating the logger", LOGGER_PREFIX);
+        Logger logger = loggerContextExtension
+                .getLoggerContext()
+                .getLogger(RedisAppenderPubSubTest.class);
+
+        final JedisPubSubTestImpl pubsub = new JedisPubSubTestImpl();
+        executor.execute(() -> redisClientExtension.getClient().subscribe(pubsub, redisKey));
```

Comment:
Why do we need to invoke `subscribe()` in the background? Why doesn't calling `subscribe()` in the current thread work for the purposes of the test?

Reply:
Because it is a blocking call and the test does not progress, we need a thread to be subscribed for the messages to be sent, but not the current one.